nodeJS/express/controllers.md

```diff
@@ -0,0 +1,451 @@
+- Know basic file structure
+
+### Handling responses
+
+When it comes to sending responses from your controllers, you have several methods at your disposal. Let's explore some of the commonly used methods and their use cases.
+
+- [res.send](https://expressjs.com/en/api.html#res.send) - A general-purpose method for sending a response, it is flexible with what data you can send since it will automatically set the `Content-Type` header based on what data you pass, and yes including setting to an `application/json`.
+- [res.json](https://expressjs.com/en/api.html#res.json) - If you want to send a JSON response, `res.json` is the way to go. It automatically sets the `Content-Type` header to `application/json` and sends the data as JSON. We will cover the use-case of this more later.
+- [res.redirect](https://expressjs.com/en/api.html#res.redirect) - When you want to redirect the client to a different URL, this method allows for that capability.
+- [res.render](https://expressjs.com/en/api.html#res.render) - If you're using a template engine that you will setup in a later lesson, `res.render` allows you to render a view template and send the rendered HTML as the response.
+
+There is also a useful method that you can use to set the status code manually.
+
+- [res.status](https://expressjs.com/en/api.html#res.status) - You can also chain other methods through this (e.g. `res.status(200).send(...)` but note that you can't do `res.send(...).status(200)`). This method does not end the request-response cycle. And also `200` is already the default so you wouldn't necessarily need this for `200`, but for consistency, we'll keep using it.controller
+
+<div class="lesson-note" markdown="1">
+
+#### res.send and res.json
+
+If `res.send` automatically sets the `Content-Type` based on the data passed. Why still use `res.json`? It's because `res.json` enforces JSON and will automatically convert non-object to JSON but `res.send` will not. But `res.json` is just a convenient method that also internally calls `res.send` but it handles anything JSON related. `res.send` also calls `res.json` internally but you will only see JSON type of `Content-Type` with booleans and objects.
+
+So for convenience it's more appropriate to use `res.json` instead of `res.send`, and if you are sending JSON then you might as well go for a method that is *named "json"*  :)
+
+</div>
+
+We also need to take note that these response methods only ends the request-response cycle, it does not end the function execution. So for example if you somehow do this:
+
+```javascript
+app.use((req, res) => {
+  res.send("Hello"); // This works and this ends the request-response cycle
+
+  console.log('will still run!!'); // however it does not exit the function so this will still continue to run
+  res.send("Bye"); // This will not work but it will throw an error that you cannot send again after sending it to the client already
+})
+```
```

Comment:
````suggestion
```javascript
app.use((req, res) => {
  // This works and this ends the request-response cycle
  res.send("Hello");

  // However, it does not exit the function so this will still run
  console.log('will still run!!');
  
  // This will then throw an error that you cannot send again after sending to the client already
  res.send("Bye"); 
});
```
````
Mainly formatting to keep comments above the respective lines (since we're not using a comment to show an output)

Reply:
Ahh yeah I forgot that comments next to code are usually outputs